File: SourceCode/AgroParallel/Web/AgroParallel.WebUI/wwwroot/js/qx-agro.js
// ============================================================================
// qx-agro.js — conversión de telemetría cruda (pps) a unidades agronómicas.
//
// El operario NO ve pps. Ve, según el tipo de motor:
//   · motor de semilla  (unidad_dosis === 'sem_m')  → sem/m, sem/10m, sem/ha
//   · motor fertilizante (unidad_dosis === 'kg_ha') → kg/ha
//
// Fórmulas inversas verificadas contra QuantiXMotorBridge.cs (la dosis objetivo
// se convierte a pps en el bridge; acá hacemos el camino inverso con el pps real).
// La calibración (semillas_vuelta/dientes_engranaje, meter_cal) es la que define
// cuántas semillas o gramos tira cada pulso — NO el peso de la semilla.
// ============================================================================
(function (global) {
  'use strict';

  // motor: config del motor (unidad_dosis, semillas_vuelta, dientes_engranaje,
  //        meter_cal, cortes).
  // ppsReal: pulsos/seg medidos por el encoder del motor.
  // ctx: { speedKmh, toolWidthM, rowSpacingM }.
  // Devuelve { mode:'sem'|'kg', valid, sem_m, sem_10m, sem_ha, kg_ha }.
  function units(motor, ppsReal, ctx) {
    motor = motor || {};
    ctx = ctx || {};
    var vMs = (ctx.speedKmh || 0) / 3.6;
    var pps = ppsReal || 0;

    if (motor.unidad_dosis === 'sem_m') {
      // semillas/pulso = semillas por vuelta / pulsos del encoder por vuelta.
      var ppvSem = (motor.dientes_engranaje > 0) ? motor.dientes_engranaje : 24;
      var semPorPulso = (motor.semillas_vuelta || 0) / ppvSem;
      var surcos = (motor.cortes && motor.cortes.length) ? motor.cortes.length : 1;
      var spacing = (ctx.rowSpacingM > 0) ? ctx.rowSpacingM : 0.525;
      var semM = (vMs > 0) ? (pps * semPorPulso / (vMs * surcos)) : 0;
      return {
        mode: 'sem',
        valid: vMs > 0,
        sem_m: semM,
        sem_10m: semM * 10,
        sem_ha: (spacing > 0) ? (semM * 10000 / spacing) : 0,
        kg_ha: 0
      };
    }

    // kg/ha: meter_cal = GRAMOS por pulso. Ancho = ancho total de labor (eje sólido).
    var meterCal = (typeof motor.meter_cal === 'number') ? motor.meter_cal : 50;
    var ancho = (ctx.toolWidthM > 0) ? ctx.toolWidthM : 0;
    var kgHa = (vMs > 0 && ancho > 0) ? (pps * meterCal * 10 / (ancho * vMs)) : 0;
    return {
      mode: 'kg',
      valid: vMs > 0 && ancho > 0,
      sem_m: 0, sem_10m: 0, sem_ha: 0,
      kg_ha: kgHa
    };
  }

  // Valor + unidad "principal" de un motor (lo que va en cajita grande / objetivo).
  function primary(u) {
    if (!u) return { v: '\u2014', u: '' };
    if (u.mode === 'sem') {
      return { v: u.valid ? u.sem_m.toFixed(1) : '\u2014', u: 'sem/m' };
    }
    return { v: u.valid ? u.kg_ha.toFixed(1) : '\u2014', u: 'kg/ha' };
  }

  // Línea compacta para listas/overlay (sem → sem/m + sem/ha; kg → kg/ha).
  function label(u) {
    if (!u) return '\u2014';
    if (!u.valid) return 'sin velocidad';
    if (u.mode === 'sem') {
      return u.sem_m.toFixed(1) + ' sem/m \xb7 ' + Math.round(u.sem_ha) + ' sem/ha';
    }
    return u.kg_ha.toFixed(1) + ' kg/ha';
  }

  // Construye el ctx a partir del estado del implemento central + velocidad PilotX.
  function ctxFrom(implCentral, speedKmh) {
    implCentral = implCentral || {};
    return {
      speedKmh: speedKmh || 0,
      toolWidthM: implCentral.ancho_total_m || 0,
      rowSpacingM: implCentral.distancia_entre_surcos_m || 0.525
    };
  }

  global.qxAgro = { units: units, primary: primary, label: label, ctxFrom: ctxFrom };
})(window);

File: SourceCode/PilotX.UI/Services/QxAgro.cs
// ============================================================================
// QxAgro.cs — port 1:1 de wwwroot/js/qx-agro.js.
//
// QUÉ QUEDÓ NATIVO: toda la conversión de telemetría cruda (pps) a unidades
// agronómicas que usa el editor nativo de QuantiX (QuantiXEditorPanel).
// QUÉ SIGUE EN HTML: qx-agro.js sigue existiendo tal cual porque la página
// pages/quantix.html la sigue usando el celular/PWA (regla del repo: las
// páginas HTML no se borran).
//
// POR QUÉ DUPLICADO: las fórmulas ya viven en QuantiXMotorBridge.cs (camino
// dosis→pps) y en qx-agro.js (camino inverso, para pantalla). Esta es la
// tercera copia, la del panel nativo. Si se toca el criterio en el bridge,
// hay que tocarlo acá también — ver aviso cruzado en QuantiXMotorBridge.cs.
//
// REGLA DE PRODUCTO: el operario JAMÁS ve pps. Ve sem/m, sem/ha o kg/ha, y
// rpm para saber si el motor responde.
// ============================================================================

using System;
using System.Globalization;

namespace PilotX.Desktop.Services;

/// <summary>Contexto agronómico de la máquina: velocidad, ancho de labor y
/// distancia entre hileras. Sale del implemento central pisado por
/// /api/tool (PilotX es la fuente de verdad del ancho).</summary>
public sealed class QxAgroCtx
{
    public double SpeedKmh   { get; set; }
    public double ToolWidthM { get; set; }
    public double RowSpacingM { get; set; } = 0.525;
}

/// <summary>Resultado de la conversión. Mode: "sem" o "kg".</summary>
public sealed class QxAgroUnits
{
    public string Mode  { get; set; } = "kg";
    public bool   Valid { get; set; }
    public double SemM  { get; set; }
    public double Sem10m { get; set; }
    public double SemHa { get; set; }
    public double KgHa  { get; set; }
}

public static class QxAgro
{
    /// <summary>Pulsos por vuelta del motor (dientes_engranaje). 24 es el
    /// default histórico del JS cuando el motor no lo trae.</summary>
    private static double Ppr(QxMotorConfig? m)
        => (m != null && m.DientesEngranaje > 0) ? m.DientesEngranaje : 24.0;

    /// <summary>pps → rpm del eje del dosificador.</summary>
    public static double PpsToRpm(QxMotorConfig? m, double pps)
    {
        double ppr = Ppr(m);
        return ppr > 0 ? (pps / ppr * 60.0) : 0.0;
    }

    public static QxAgroCtx CtxFrom(QxImplemento? impl, double speedKmh)
    {
        double spacing = (impl != null && impl.DistanciaEntreSurcosM > 0)
            ? impl.DistanciaEntreSurcosM : 0.525;
        return new QxAgroCtx
        {
            SpeedKmh    = speedKmh,
            ToolWidthM  = impl?.AnchoTotalM ?? 0,
            RowSpacingM = spacing,
        };
    }

    public static QxAgroUnits Units(QxMotorConfig? motor, double ppsReal, QxAgroCtx? ctx)
    {
        ctx ??= new QxAgroCtx();
        double vMs = ctx.SpeedKmh / 3.6;
        double pps = ppsReal;

        if (motor != null && string.Equals(motor.UnidadDosis, "sem_m", StringComparison.Ordinal))
        {
            // semillas/pulso = semillas por vuelta / pulsos del encoder por vuelta.
            double semPorPulso = motor.SemillasVuelta / Ppr(motor);
            int surcos = (motor.Cortes != null && motor.Cortes.Count > 0) ? motor.Cortes.Count : 1;
            double spacing = ctx.RowSpacingM > 0 ? ctx.RowSpacingM : 0.525;
            double semM = vMs > 0 ? (pps * semPorPulso / (vMs * surcos)) : 0.0;
            return new QxAgroUnits
            {
                Mode   = "sem",
                Valid  = vMs > 0,
                SemM   = semM,
                Sem10m = semM * 10.0,
                SemHa  = spacing > 0 ? (semM * 10000.0 / spacing) : 0.0,
                KgHa   = 0,
            };
        }

        // kg/ha: meter_cal = GRAMOS por pulso. Ancho = ancho total de labor.
        double meterCal = motor != null ? motor.MeterCal : 50.0;
        double ancho = ctx.ToolWidthM > 0 ? ctx.ToolWidthM : 0.0;
        double kgHa = (vMs > 0 && ancho > 0) ? (pps * meterCal * 10.0 / (ancho * vMs)) : 0.0;
        return new QxAgroUnits { Mode = "kg", Valid = vMs > 0 && ancho > 0, KgHa = kgHa };
    }

    /// <summary>Valor grande + unidad ("3.5" / "sem/m").</summary>
    public static (string V, string U) Primary(QxAgroUnits? u)
    {
        if (u == null) return ("—", "");
        if (u.Mode == "sem")
            return (u.Valid ? u.SemM.ToString("0.0", CultureInfo.InvariantCulture) : "—", "sem/m");
        return (u.Valid ? u.KgHa.ToString("0.0", CultureInfo.InvariantCulture) : "—", "kg/ha");
    }

    /// <summary>Línea compacta de lista/overlay.</summary>
    public static string Label(QxAgroUnits? u)
    {
        if (u == null) return "—";
        if (!u.Valid) return "sin velocidad";
        if (u.Mode == "sem")
            return u.SemM.ToString("0.0", CultureInfo.InvariantCulture) + " sem/m · "
                 + Math.Round(u.SemHa).ToString("0", CultureInfo.InvariantCulture) + " sem/ha";
        return u.KgHa.ToString("0.0", CultureInfo.InvariantCulture) + " kg/ha";
    }
}
